docs/proposals/component-registration.md

```diff
@@ -0,0 +1,363 @@
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/component-registration.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Proposal: Component Registration
+
+## Table of Contents
+
+<!-- BEGIN MUNGE: GENERATED_TOC -->
+
+- [Proposal: Component Registration](#proposal-component-registration)
+  - [Table of Contents](#table-of-contents)
+  - [Abstract](#abstract)
+  - [Use Cases](#use-cases)
+    - [Deployment](#deployment)
+    - [Debugging](#debugging)
+    - [Notification](#notification)
+    - [Reporting](#reporting)
+    - [Self-Healing](#self-healing)
+  - [Design](#design)
+    - [API Changes](#api-changes)
+      - [Component API](#component-api)
+      - [ComponentStatuses API](#componentstatuses-api)
+    - [Kubectl (CLI)](#kubectl-cli)
+    - [Storage](#storage)
+    - [Status Updates](#status-updates)
+  - [History](#history)
+  - [Example](#example)
+  - [Next Steps](#next-steps)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+## Abstract
+
+Kubernetes was designed to be extensible from the start, but one factor currently limits that capability: the hardcoding of what defines a Kubernetes cluster. To reach enterprise-class capabilities (high availability, notifications, self-healing, reporting, and maintainability) Kubernetes first needs to support dynamic component registration and readiness probes.
+
+
+## Use Cases
+
+This proposal primarily covers features that can be achieved internally by Kubernetes itself, but it's clear that there are also some desirable features that must be provided by external deployment, management, or monitoring systems. These external systems, however, require some features within Kubernetes to enable them. So while not all of these use cases can be solved by Kubernetes alone, we do want to make them possible.
+
+### Deployment
+
+As a k8s operator, I want a kubectl command that allows me to validate that the cluster has been deployed and is fully functional.
+
+As a k8s operator, I want a kubectl command that allows me to check which components are deployed/installed on a specific cluster in order to infer which features it supports.
+
+Note: Inferring features from components isn't very useful with vanilla k8s, but becomes more useful as 3rd party or optional components are added (e.g. OpenShift) and/or if components-controller is split into multiple components.
+
+### Debugging
+
+As a k8s operator, I want a single endpoint (and kubectl command) where I can see the current condition of the cluster so that I can debug observed failures (dropped connections, unresponsiveness, etc.).
+
+Note: The endpoint should contain cause where possible, but at least the current condition of the components.
```

Comment:
This can start simple but at some point I'd love to see it be a more fully featured "deep systems check" type of thing.  Example would be `brew doctor` for homebrew.